rafthttp: support sending v3 snapshot message

yichengq opened the pull request:
Use pipelineSnapshot to send v3 snapshot message. It puts v3 snapshot
message bytes into request body, then sends it to the target peer.
When it receives http.StatusNoContent, it knows the message has been
sent successfully.

As receiver, snapHandler receives full v3 snapshot message, saves snapshot
, processes the raft snapshot message, then respond with
http.StatusNoContent.

I have manually tested it.

follow #3648 
for #3549

xiang90 commented:
I cannot understand the design you made here. Why do we call the snapshot sender a pipeline? If I understand the code correctly, after it sends out one snapshot request the other side will close the HTTP request. I cannot understand why it is a pipeline at all. :(

yichengq commented:
I thought making it align with pipeline could make code easier to read. It
does send http request for each message, and close it after sending. Based
on your description, it doesnt match its name. I will rename it back to
snapshotSender.

On Mon, Oct 12, 2015 at 18:53 Xiang Li <notifications@github.com> wrote:

> I cannot understand the design you made here. Why do we call the snapshot
> sender a pipeline? If I understand the code correctly, after it sends out
> one snapshot request the other side will close the HTTP request. I cannot
> understand why it is a pipeline at all. :(
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/coreos/etcd/pull/3666#issuecomment-147571714>.
>

xiang90 commented:
@yichengq The logic also does not make sense to me. I do not understand why do we need a dedicated go routine to wait for the snapMsg.

yichengq commented:
@xiang90 Updated with no dedicated goroutine to wait for the snapMsg. Originally I wanted to use an long-time entity to manage snapMsg send, but it became heavy. Remove it to make the logic simple.

xiang90 commented:
Can you explain to me what will happen when:

1. the snapshot sender dies without closing underlying TCP connection?

2. the snapshot receiver dies without closing underlying TCP connection?

3. If the snapshot is large and takes a long time, is there anything will be affected?

4. If one snapshot sending fails, how the does retry mechanism work?

5. what happens if a leader wants to send 2 snapshot concurrently to 2 followers?

xiang90 commented:
I am asking since I cannot find out the handling logic from code or any related timeout setting or comments.

yichengq commented:
> the snapshot sender dies without closing underlying TCP connection?

I think the receiver keeps waiting for the request body, and stops the handler when it detects TCP connection is down due to TCP keepalive mechanism. If another leader is elected in this period, and send it the snapshot with the same index, it still could save it to disk and recover from there. So I think it is fine.

> the snapshot receiver dies without closing underlying TCP connection?

If receiver dies when it is sending request body, receiver should be notified soon because TCP connection cannot write anything out.

If receiver dies when it is waiting for response, receiver needs to wait for TCP keepalive to notice it about the connection down. **This is something missing now, I will improve.**

> If the snapshot is large and takes a long time, is there anything will be affected?

It could occupy all resources and even affect heartbeat between leader and followers. My plan for this is to test and fix it after this PR.

> If one snapshot sending fails, how the does retry mechanism work?

If one snapshot fails, rafthttp.snapshotSender will report SnapshotFailure to raft state machine. raft state machine change its progress to `probe` and pause a while. Then raft state machines makes a new snapshot and send it to the follower again.

> what happens if a leader wants to send 2 snapshot concurrently to 2 followers?

The first snapshot will be made successfully, and send out. The second snapshot will be blocked because snapshot cannot be made due to `ErrSnapshotTemporarilyUnavailable`. After the first snapshot is sent out, the second snapshot could be made and sent out.

I would try to address these questions through more comments.

xiang90 commented:
> If receiver dies when it is waiting for response, receiver needs to wait for TCP keepalive to notice it about the connection down. This is something missing now, I will improve.

What is your plan? The affects my 5 question a lot.

> It could occupy all resources and even affect heartbeat between leader and followers. My plan for this is to test and fix it after this PR.

What is the plan?

> The second snapshot will be blocked 

What do you mean by blocking? From the code, I think the snapshot sending will be canceled rather than blocked.

yichengq commented:
My draft plan for `receiver dies when sender is waiting for response` is to let snapshotSender cancel the request if it has waited for a long time(~5s) after it has write out the full request body.

My plan for `large snapshot sent` is to send the snapshot block by block, and rest some time between each block sent to let main loop work.

> What do you mean by blocking? From the code, I think the snapshot sending will be canceled rather than blocked.

When raft state machine wants to send snapshot message to the second member at that time, it will receive `ErrSnapshotTemporarilyUnavailable` and cancel the snapshot sending. After some time, it would try to send snapshot message again. `blocking` is inaccurate, it is `keep retrying`.

xiang90 commented:
@yichengq Can you please add TODOs into the current code? So that I can properly review the code without worrying about these issues?

yichengq commented:
TODO added. I am adding comments to address the above 5 questions now.

yichengq commented:
comments for questions added. Ready for the review.

xiang90 commented on rafthttp/http.go:
```diff
@@ -122,6 +134,93 @@ func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+// the handler will keep waiting for the request body until TCP keepalive
+// finds out that the connection is broken after several minutes.
+// This is acceptable because
+// 1. following snapshot messages could still be received and processed
```
what is `the following snapshot`? why it can still be received and processed?

yichengq replied:
`snapshot messages sent through other TCP connections could still be received and processed.`

xiang90 commented on rafthttp/http.go:
```diff
@@ -122,6 +134,93 @@ func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+		return
+	}
+
+	if err := checkVersionCompability(r.Header.Get("X-Server-From"), serverVersion(r.Header), minClusterVersion(r.Header)); err != nil {
```
it seems that line159->line173 are duplicate code. we already have them somewhere else in the codebase. we should make it a function.

yichengq replied:
will try.

xiang90 commented on rafthttp/http.go:
```diff
@@ -122,6 +134,93 @@ func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	var l uint64
+	// read the length of raft message
+	if err := binary.Read(r.Body, binary.BigEndian, &l); err != nil {
```
do we have a decoder somewhere to decode a raft msg?

yichengq replied:
no. We do decode raft msg in some stream types, but they have their own wire formats.

xiang90 replied:
what is their format? Not len | binary_proto_msg?

yichengq replied:
The format is `len | binary_proto_msg`. I check the encoder and decoder again, and it fits the format here. I resisted to reuse them before because encodoer/decoder is originally designed for stream, but it is no harm to use them when reading over this again.

xiang90 replied:
> and it fits the format here
So `but they have their own wire formats` is incorrect?

> encodoer/decoder is originally designed for stream

No, the original design is not. The decoder/encoder should be designed just for encoding and decoding. Just like JSON.Marahal is not designed for HTTP.

>  no harm to use them when reading over this again

It is not only no harm, but brings us a lot of benefits.

yichengq replied:
Agree. It is unreasonable that I abstracted it but resisted to reuse it.

xiang90 commented on rafthttp/http.go:
```diff
@@ -122,6 +134,93 @@ func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	if err := h.r.Process(context.TODO(), m); err != nil {
+		switch v := err.(type) {
+		case writerToResponse:
```
raft will generate writerToResponse type error?

yichengq replied:
Process is etcdserver's function that wraps raft.Step and does some checks, and it will generate writerToResponse error.

xiang90 replied:
As we discussed offline, we should write comment on these unclear parts.

xiang90 commented on rafthttp/http.go:
```diff
@@ -122,6 +134,93 @@ func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+			v.WriteTo(w)
+		default:
+			plog.Warningf("failed to process raft message (%v)", err)
+			http.Error(w, "error processing raft message", http.StatusInternalServerError)
```
can we attach the original error like line 214?

yichengq replied:
sure.

xiang90 commented on rafthttp/peer.go:
```diff
@@ -76,8 +77,8 @@ type Peer interface {
-// Each peer has two underlying mechanisms to send out a message: stream and
-// pipeline.
+// Each peer has three underlying mechanisms to send out a message: stream,
+// pipeline and snapshot sender.
```
snapshot sender is not a mechanism.

yichengq replied:
I don't understand this one. snapshot sender is the way to send out snapshot message.

xiang90 replied:
snapshot sender is not a way, a mechanism or a method. It is an entity.

yichengq replied:
I see. will try to improve it.

xiang90 commented on rafthttp/peer.go:
```diff
@@ -85,16 +86,20 @@ type Peer interface {
 // to the remote follower node.
 // A pipeline is a series of http clients that send http requests to the remote.
 // It is only used when the stream has not been established.
+// Snapshot sender is the only mechanism that support sending out
```
A sender is not a mechanism.

xiang90 commented on rafthttp/snapshot_sender.go:
```diff
@@ -0,0 +1,205 @@
+		// new snapshot message.
+		s.r.ReportSnapshot(m.To, raft.SnapshotFailure)
+		plog.Infof("snapshot [index: %d, to: %s] failed to be sent out (%v)", m.Snapshot.Metadata.Index, types.ID(m.To), err)
+	} else {
```
return after line 86. remove else.

xiang90 commented on rafthttp/snapshot_sender.go:
```diff
@@ -0,0 +1,205 @@
+		// machine knows about it, it would pause a while and retry sending
+		// new snapshot message.
+		s.r.ReportSnapshot(m.To, raft.SnapshotFailure)
+		plog.Infof("snapshot [index: %d, to: %s] failed to be sent out (%v)", m.Snapshot.Metadata.Index, types.ID(m.To), err)
```
This should be an warning rather than info

yichengq replied:
I think I will print it out based on the peer status.

xiang90 commented on rafthttp/snapshot_sender.go:
```diff
@@ -0,0 +1,205 @@
+			return
+		}
+		body, err := ioutil.ReadAll(resp.Body)
+		defer resp.Body.Close()
```
why not just close? what does defer do here?

yichengq replied:
it works the same way without defer. will remove.

xiang90 replied:
It does not work the same way. It is much better without the unnecessary defer. We should never misuse defer.

xiang90 commented on rafthttp/snapshot_sender.go:
```diff
@@ -0,0 +1,205 @@
+		resp, body = r.resp, r.body
+	}
+
+	switch resp.StatusCode {
```
This is also a duplicate code block. Should there be a checkStatusCode func?

yichengq replied:
We don't have it today. I could make it a helper function.

xiang90 replied:
We should add it today. We should not keep on pushing in duplicate code. There are already a bunch of them in rafthttp pkg. And they have never been cleaned up. :(

yichengq replied:
> And they have never been cleaned up. :(

After I work on it, i find it is definitely true. I will follow up with cleanup after the snapshot stuffs.

I just read through the quite similar code block. I would put use shared functions for them if there is few changes. The left part will be addressed when I clean it up.

xiang90 commented on rafthttp/snapshot_sender.go:
```diff
@@ -0,0 +1,205 @@
+}
+
+// createSnapBody creates the request body for the given raft snapshot message.
+// The returned body should always be closed when it is no longer used.
```
Callers should close body when done reading from it.

xiang90 commented on rafthttp/snapshot_sender.go:
```diff
@@ -0,0 +1,205 @@
+		plog.Panicf("body write error (%v)", err)
+	}
+	// get snapshot
+	// TODO: the snapshot could be large and eat up all resources when writing
```
why do you put the comment here? it is unrelated in my opinion.

yichengq replied:
i should put it at the place to send the request. i originally thought it affects how to build the body.

xiang90 replied:
Does it affect how to build the body?

yichengq replied:
Basically the new Reader should be a kind-of `PartTimeReader` upon original `io.Reader`. It is better to add this layer in send() function. So I will remove it from building the body.

xiang90 replied:
PartTimeReader is a terrible name btw.

yichengq replied:
@xiang90 Just a joke. :wink:

xiang90 commented on rafthttp/transport.go:
```diff
@@ -307,3 +329,23 @@ func (t *Transport) Resume() {
 	}
 }
+
+// snapshotStore is the temporary store of snapshot. Caller could put one
```
is this store temporary?

xiang90 replied:
Caller can

yichengq replied:
Yes. it expects user get the snapshot out in a short time.

xiang90 replied:
Why the store is a temporary? When do you create it? When do you destroy it?

yichengq replied:
i see your point. the store is not a temporary one and it exists in the whole life cycle of Transport. Calling it `store of temporary snapshot` is weird too because the snapshot has its own life cycle. I guess I will just delete it.

xiang90 commented on rafthttp/transport.go:
```diff
@@ -307,3 +329,23 @@ func (t *Transport) Resume() {
+
+// snapshotStore is the temporary store of snapshot. Caller could put one
+// snapshot into the store, and get it later.
+// snapshotStore stores at most one snapshot at a time.
```
what if the user puts more than once?

yichengq replied:
it stores the new one then. I should let it panic here to let the comment always be true.

yichengq commented on rafthttp/pipeline.go:
```diff
@@ -170,31 +156,9 @@ func (p *pipeline) post(data []byte) (err error) {
-	}
+	err = checkPostResponse(resp, b, req, p.to)
+	reportCritialErr(err, p.errorc)
+	return err
```
It reports the critical error here to keep the original logic. Or it may need changes to related tests. Will clean it up in follow-ups

yichengq commented on rafthttp/pipeline.go:
```diff
@@ -170,31 +156,12 @@ func (p *pipeline) post(data []byte) (err error) {
 		return nil
-	default:
-		return fmt.Errorf("unexpected http status %s while posting to %q", http.StatusText(resp.StatusCode), req.URL.String())
 	}
```
It reports the critical error here to keep the original logic. Or it may need changes to related tests. Will clean it up in follow-ups

yichengq commented:
Addressed all comments. PTAL.

yichengq commented:
Addressed comments about function names. Hope they look better this time.

yichengq commented:
Refine the names. PTAL

yichengq commented:
Address previous round of comments. Check it to see whether it is good.

xiang90 commented:
Have you manually tested the failure situation? There is basically no test. Do we have a plan to add more tests?

yichengq commented:
> Have you manually tested the failure situation?

Haven't yet. It still misses one PR to make snapshot send work, another PR to make large snapshot send work. When I manually tested it about normal situation, I added some hack to make it work. I would like to test failure cases after the missing one is added.

BTW, list some possible failure cases that come into my head now: snapshot sender dies when sending, snapshot receiver dies when receiving, sending to two followers.

> Do we have a plan to add more tests?

Definitely. You will see me send PR after getting all these done. I have added one TODO in https://github.com/coreos/etcd/issues/3550, and it would need integration tests.

xiang90 commented:
Can you please squash all the commits into one? Then I will take a final look. It is pretty close.

yichengq commented:
squash done.

xiang90 commented:
Is the commit message correct? 
```
Use pipelineSnapshot to send v3 snapshot message. It puts v3 snapshot
message bytes into request body, then sends it to the target peer.
When it receives http.StatusNoContent, it knows the message has been
sent successfully.
```

xiang90 commented on rafthttp/http.go:
```diff
@@ -122,6 +123,75 @@ func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+		http.Error(w, msg, http.StatusInternalServerError)
+		return
+	}
+
```
can we log something out here to inform that we received a snapshot?

yichengq replied:
sure. added line to print like this:
```
2015-10-13 22:59:57.296356 I | rafthttp: received snapshot [index: 211535, from: 8211f1d0f64f3269] successfully
```

xiang90 replied:
received and saved snapshot ...

xiang90 commented:
LGTM

yichengq commented:
> Is the commit message correct?

No. Have updated to match the current situation.

xiang90 commented:
LGTM again.

yichengq commented:
waiting for CI pass.
